test/model-builder.test.js

```diff
@@ -77,6 +77,109 @@ describe('ModelBuilder', () => {
       });
     });
 
+    describe('model with nested properties as embedded model', () => {
+      let Address, Person;
+      beforeEach('Define models', () => {
+        Address = builder.define('Address', {
+          street: {type: 'string'},
+          number: {type: 'number'},
+        });
+        Person = builder.define('Person', {
+          name: {type: 'string'},
+          address: {type: 'Address'},
+          other: {type: 'object'},
+        });
+      });
+      it('should properly add the __parent relationship when instantiating parent model', () => {
+        const person = new Person({
+          name: 'Mitsos',
+          address: {street: 'kopria', number: 11},
+        });
+        person.should.have.propertyByPath('address', '__parent').which.equals(person);
+      });
+      it('should add _parent property when setting embedded model after instantiation', () => {
+        const person = new Person({
+          name: 'Mitsos',
+        });
+        person.address = {street: 'kopria', number: 11};
+        person.should.have.propertyByPath('address', '__parent').which.equals(person);
+      });
+      it('should handle nullish embedded property values', () => {
+        const person = new Person({
+          name: 'Mitsos',
+          address: null,
+        });
+        person.should.have.property('address').which.equals(null);
+      });
+      it('should properly re-set the parent property when moving a child instance to an other parent', () => {
+        const person1 = new Person({
+          name: 'Mitsos',
+          address: {street: 'kopria', number: 11},
+        });
+        const {address} = person1;
+        address.should.be.instanceof(Address).and.have.property('__parent').which.equals(person1);
+        const person2 = new Person({
+          name: 'Allos',
+          address,
+        });
+        address.should.have.property('__parent').which.equals(person2);
```

Comment:
Hmm, I find this suspicious. IIUC, the following assertion would be also true:

```ts
person1.address.should.have.property('__parent').which.equals(person2);
```

In other words, it's possible to have an embedded value that's pointing to a wrong parent.

**I think this is a major design issue we need to find a solution for.**

Can we throw an error when an embedded value is re-assigned from one parent to another? What other options do we have?

Reply:
You are right about this one. I could throw an error, but this would create error in user programs where they assigned whole child models to other parent instances. Since the `__parent` is statically stored on the child model, it can only reference one parent model instance. If you are ok with the possible error case I will implement it

Reply:
Personally, I prefer solutions that handling edge cases in a robust and least-surprising way, thus I think it's best to throw an error when a child model is reassigned to a different parent. 

However, such operation may be legitimate, therefore we need an easy way how to do the reassignment correctly. I think the recipe has two steps:
- Remove the child from the original parent, this should reset `__parent` reference(s) to `undefined`. Is there any relational API like `unlink` that can perform this step in a safe way? Something along the lines of `await category.products.unlink(product)` to remove `product` instance from the parent `category` instance.
- Attach the child to the new parent, this will set `_parent` reference(s) to the new parent.

I am concerned about breaking existing code that may be reassigning child models between parents and not caring about the `__parent` reference at all. If we start throwing an error then we will break existing applications. To preserve backwards compatibility, I am proposing to add `__parent` reference only when a feature flag is enabled. 

There are different places where this feature flag can be specified. I think it would be great to have one place to enable it for many models - perhaps via the ModelBuilder instance that's shared by all datasources in a single LB3 app? Other alternatives: datasource config (`server/datasources.json`), relation definition (`common/models/category.json`, `relations.products` object).

Reply:
I agree about the **feature flag** option. It could be in the **datasources.json**.

As for the re-assignemnt error, requiring the extra step of unlinking I think is not necessary, because after all the new parent would still be applied on the original child model instance which will create a problem if it is still attached to the original parent model.

Maybe we could keep the current behavior of re-assignment and just issue a console warning message that an actual child instance is being re-assigned???

Reply:
Child re-assignment now issues a WARNING in console instructing the user to avoid such operations. Also complete functionality is enabled only when setting the feature flag `modelBuilder.settings.parentRef` to `true`